src/charm.py

```diff
@@ -37,24 +45,97 @@ def __init__(self, *args):
+        # new config file to the workload container
+        prometheus_config = self._prometheus_config()
+        config_hash = str(hashlib.md5(str(prometheus_config).encode('utf-8')))
+        if not self._stored.prometheus_config_hash == config_hash:
+            self._stored.prometheus_config_hash = config_hash
+            container.push(PROMETHEUS_CONFIG, prometheus_config)
+            logger.info("Pushed new configuration")
+
+        # setup the workload (Prometheus) container and its services
+        layer = self._prometheus_layer()
+        plan = container.get_plan()
+        if plan.services != layer["services"]:
+            container.add_layer("prometheus", layer, combine=True)
+
+            if container.get_service("prometheus").is_running():
+                container.stop("prometheus")
+
+            container.start("prometheus")
+            logger.info("Restarted prometheus container")
+
+        self.app.status = ActiveStatus()
+        self.unit.status = ActiveStatus()
+
+    def _on_update_status(self, event):
+        """Check status of Prometheus server.
+
+        Status of the Prometheus services is checked by querying
+        Prometheus for its version information. If Prometheus responds
+        with valid information, its active status is recorded.
+        """
+        provided = {'prometheus': self.version}
+        if provided['prometheus']:
+            logger.debug("Prometheus provider is available")
+            logger.debug("Providing : {}".format(provided))
+            if not self._stored.provider_ready:
+                self._stored.provider_ready = True
 
     def _on_stop(self, _):
-        """Mark unit is inactive
+        """Mark unit is inactive.
+
+        All units of the charm are set to maintenance status before
+        termination.
         """
         self.unit.status = MaintenanceStatus('Pod is terminating.')
 
     def _on_grafana_changed(self, event):
-        """Provide Grafana with data source information
+        """Provide Grafana with data source information.
+
+        Grafana needs to know the port and name of an application in order
+        to form a relation with it. Hence this information is provided here.
         """
         event.relation.data[self.unit]['port'] = str(self.model.config['port'])
         event.relation.data[self.unit]['source-type'] = 'prometheus'
+        event.relation.data[self.unit]['private-address'] = str(
+            self.model.get_binding(event.relation).network.bind_address
+        )
 
     def _on_alertmanager_changed(self, event):
-        """Set an alertmanager configuation
+        """Set an alertmanager configuration.
+
+        In response to any changes in relations with Alertmanager,
+        the list of currently available Alertmanagers is updated,
+        a new Prometheus configuration set and Prometheus is
+        restarted.
         """
         if not self.unit.is_leader():
             return
```

Comment:
what do you think about a `@leader_only` decorator?

Reply:
I hesitate to speculate :-) . However since you bring up the issue.  Would be the use case for such a decorator ? What would be the behaviour of such a decorator ? Would it immediately return if the current unit is not leader ? This can be done for the present. However the alertmanager event handler methods in this charm are very rudimentary. For example there is no logic there for handling forwarding rules from applications to prometheus to alertmanager. Even though this behaviour has not yet been discussed, agreed, and specified, it would be the next logical step to do once our charm libraries and pebble migration is complete. Would you suppose these altermanager event handlers would have no role to play if the unit is not leader once we add the logic for handling alerting rules and such ?.

Reply:
I see this exact `is_leader` preamble in many places in many repos, that's why I ask :)
 Don't know yet about alerting rules.

Reply:
One can make a use case for `@leader_only` to be a utility in the Operator Framework. It would not reduce the duplication of code much if each "repo" tried to implement its own `@leader_only`. The use case may be stronger if you can think of a repertoire of such higher order functions. At present leader only seems to be the only reasonable one I have come across.

Reply:
From my POV, I'm a fan of `@leader_only`to reduce the boilerplate

Reply:
Aye. I do think it is a reasonable suggestion too. I would rather we discuss it in stand and create a ticket for `ops` where we make a list of such utilities that would be nice too have. Besides the focus of this PR is charm library and pebble migration. I am sure that there is lot more alertmanager work to be done. I will bring it in as part of one of those commits after discussion with team in standup.

Reply:
@sed-i Let us make that github issue ticket on the Canonical Operator Framework repository, as discussed. Ping me when you are in and we can do it together. Resolving this conversation here for now.